Link Monitor

danfruehauf opened the pull request:
* Core logic (and interface)
 * Unit tests

danfruehauf commented:
This is pretty much all the logic for the monitor but doesn't include:
 * Getting all record UUIDs by querying geonetwork components (uses HTTP now and must be removed)
 * Getting record documents from geonetwork components (uses HTTP now and must be removed)
 * Disabling/enabling of records

Those are the only 3 points of interaction we should have with GN.

danfruehauf commented:
Sorry for all the code additions. This is ready for review. I'm still performing some checks but it seems to work just fine.

jkburges commented on web/src/main/java/org/fao/geonet/Geonetwork.java:
```diff
@@ -475,6 +477,16 @@ public Object start(Element config, ServiceContext context) throws Exception {
             createDBHeartBeat(context.getResourceManager(), gnContext, dbHeartBeatInitialDelay, dbHeartBeatFixedDelay);
         }
+
+        boolean linkMonitorEnabled = Boolean.parseBoolean(handlerConfig.getValue(Geonet.Config.LINK_MONITOR_ENABLED, "false"));
```
Extract method `configureLinkMonitor()`?  

Probably applies to DB heartbeat above too.  Not sure what our stand is on refactoring upstream code is, but IMO, a class with over 1000 lines is *probably* just a *little* too big :fearful:

danfruehauf replied:
This is actually the code I'm planning to try and push upstream. It's pretty difficult to try changing all of that, so I conformed with what was there before (as you said, `DBHeartBeat`).

jkburges commented on web/src/main/java/org/fao/geonet/Geonetwork.java:
```diff
@@ -558,6 +570,39 @@ private boolean checkDBWrite() {
         scheduledExecutorService.scheduleWithFixedDelay(DBHeartBeat, initialDelay, fixedDelay, TimeUnit.SECONDS);
     }
 
+    private void createLinkMonitor(final ResourceManager rm, final GeonetContext gc, String linkMonitorClass,
```
Would this be better as a factory method on `LinkMonitorInterface`?  Some funky reflection going on here, maybe that might make it difficult.

jkburges replied:
I think that this method could be re-written to be a lot more high-level (and therefore, IMO, more readable), e.g.:

```java
private void scheduleLinkMonitor(args...) {
    LinkMonitorInterface linkMonitor = LinkMonitorInterface.newInstance(args...);
    scheduledExecutorService.scheduleWithFixedDelay(linkMonitor, ...);
}
```

This assumes:

* factory method on `LinkMonitorInterface` (i.e. `newInstance()`)
* `LinkMonitorInterface` implements `Runnable` (and can therefore be passed to the executor)
* a shared executor (*probably* only need to the one thread pool between this and db heartbeat, depending on the frequency)

You can choose to do all or none of the above, or somewhere in between. :smile:

danfruehauf replied:
I think those are fair comments. Again here, I was following the `DBHeartBeat` "example". But I might change it so that `LinkMonitorInterface` implements `Runnable`, because it must.

jkburges commented on web/src/main/java/org/fao/geonet/Geonetwork.java:
```diff
@@ -558,6 +570,39 @@ private boolean checkDBWrite() {
+                }
+                // any uncaught exception would cause the scheduled execution to silently stop
+                catch(Throwable x) {
+                    logger.error("Link Monitor error: " + x.getMessage() + " This error is ignored.");
```
`logger.error("message", x)` instead of printing the stack trace to stderr?

danfruehauf replied:
Like `DBHeartBeat` (my usual excuse). We can potentially abort execution here if the user does specify that he wants link monitoring enabled and an error occurred. That be good.

jkburges commented on web/src/main/java/org/fao/geonet/kernel/search/CollectionAvailabilityFilter.java:
```diff
@@ -0,0 +1,76 @@
+        this._query = query;
+        _fieldsToLoad = Collections.singleton("_uuid");
+
+        if (serviceContext != null)
```
Is this the geonetwork way to not have braces around single liners?

danfruehauf replied:
The geonetwork way is... all over! I can get braces over all single lines, but I don't think it was consistent from what I've seen, so I allowed myself to not use braces where I saw fit. It does not conform to our coding standard, I know.

jkburges commented on web/src/main/java/org/fao/geonet/monitor/link/CheckInfo.java:
```diff
@@ -0,0 +1,16 @@
+
+import org.jdom.Element;
+
+import java.lang.reflect.InvocationTargetException;
```
Needed?

danfruehauf replied:
Nope, and removed. I gonna sweep all the classes I added for unneeded imports.

jkburges commented on web/src/main/java/org/fao/geonet/monitor/link/LinkCheckerInterface.java:
```diff
@@ -0,0 +1,9 @@
+public interface LinkCheckerInterface {
+    public boolean check(Element onlineResource);
+
+    public String toString(Element onlineResource);
```
Unusual for a `toString()` method to have a parameter in java.

jkburges commented on web/src/main/java/org/fao/geonet/monitor/link/LinkCheckerUtils.java:
```diff
@@ -0,0 +1,48 @@
+
+            LinkMonitorService.getLogger().debug(String.format("%s -> %d", url, connection.getResponseCode()));
+            if (connection.getResponseCode() != 200)
+                LinkMonitorService.getLogger().debug(String.format("URL '%s' is unavailable", url, connection.getResponseCode()));
```
It's usual for every class to have its own `Logger`.

danfruehauf replied:
I might see how to initialize the logger better, you can show me perhaps. Is it a new instance per class?

jkburges replied:
Normally,

```java
public class Wow {
  static Logger  logger = Logger.getLogger(Wow.class);
}```

So one instance shared per class.

The reason that this is better (IMO) is that then the configuration is more granular. i.e. you can be more selective about which things you want to see in the log.  I actually think this is much more of an issue for geonetwork (and geoserver) than it is for the portal, because those apps tend to have log floods, and it's pretty hard to see the bits you want.  So, tuning important.

danfruehauf replied:
Absolutely. That's my log4j/java ignorance. It will be changed.

jkburges commented on web/src/main/java/org/fao/geonet/monitor/link/LinkCheckerDefault.java:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    @Override
+    public boolean check(Element onlineResource) {
```
This method is duplicated in each of the implementation classes.  Could be extracted to a common base class.

danfruehauf replied:
It's implemented slightly different in this case. I'm actually happy with this abstraction level:
 * One class for simple links
 * One class for WMS
 * One class for WFS

WMS and WFS are a bit similar, but for the sake of clarity and readability, I prefer to have those 5-10 duplicated code lines.

jkburges replied:
I'm happy with that abstraction level too, in terms of the classes.

I respectfully and strongly disagree about the duplication though!  It's more code to understand, fix bugs in, and have diverge in behaviour, even when it shouldn't.  But I'm happy for you to keep it like this if you feel that way.

jkburges commented on web/src/main/java/org/fao/geonet/monitor/link/LinkCheckerDefault.java:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    @Override
+    public String toString(Element onlineResource) {
```
... and this one.

danfruehauf replied:
Yes, it's repeating. It happens to be that all those classes need a check url, but that's more of a coincidence. I can try extracting some of it to a base class, but what name will it have?

jkburges replied:
I was talking about `toString()` being duplicated.  Base class, maybe it would make sense to have:

```java
public abstract AbstractLinkChecker {
  public abstract boolean check(...);
  
  public String toString(...) {
    // magic happens
  }
}```

Pretty normal for an abstract class to have that in its name.

jkburges commented on web/src/main/java/org/fao/geonet/monitor/link/LinkCheckerWfs.java:
```diff
@@ -0,0 +1,35 @@
+    private static String WFS_GET_FEATURES = "?service=WFS&version=1.0.0&request=GetFeature&maxFeatures=1&outputFormat=csv&typeName=";
+
+    private String getCheckUrl(Element onlineResource) {
+        String url = LinkCheckerUtils.parseOnlineResource(onlineResource, LinkCheckerUtils.URL_XPATH);
```
This is *very* similar to `LinkCheckerWms.getCheckUrl()` - can probably be refactored.

jkburges commented on web/src/main/java/org/fao/geonet/monitor/link/LinkInfo.java:
```diff
@@ -0,0 +1,91 @@
+    private final LinkCheckerInterface linkChecker;
+
+    public LinkInfo(Element onlineResource, LinkCheckerInterface linkChecker) {
+        checkInfoList = Collections.synchronizedList(new ArrayList<CheckInfo>());
```
Why the need for a synchronized list?  In any case, we're not synchronizing on it when iterating (see http://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#synchronizedList(java.util.List)) so in case we *do* actually need it, then we've possibly got a bug.

jkburges commented on web/src/main/java/org/fao/geonet/monitor/link/LinkInfo.java:
```diff
@@ -0,0 +1,91 @@
+import java.util.Collections;
+import java.util.List;
+
+public class LinkInfo {
```
I'm finding it a bit difficult to understand the overall algorithm of checking in this class...

danfruehauf replied:
Have you read the design document? It follows it one to one.

Only one function here actually has logic.

jkburges replied:
Nope.  Where is it?

jkburges replied:
Ok, I've got a suggestion for re-writing `getStatus()` to more closely reflect what's described in [the design doc](https://github.com/aodn/internal_discussions/wiki/Geonetwork-Layer-Monitor#linkinfo-state-evaluation):

```java
public LinkMonitorService.Status getStatus() {

  if (notEnoughFreshChecks()) {
    return LinkMonitorService.UNKNOWN;
  }

  if (successfulCheckPercentage() < LinkMonitorService.percentWorkingThreshold) {
    return LinkMonitorService.FAILED;
  }

  return LinkMonitorService.WORKING;
}
```

Maybe this is not quite right - the current implementation has more `return` points than this, but IMO, it'd be really nice if the code reflected the fundamental checking algorithm more closely (with implementation details hidden down a level).  Because... the doc will die, the code will live on.

danfruehauf replied:
Yep, I'll try and change it to reflect the doc. It makes sense.

jkburges commented on web/src/main/java/org/fao/geonet/monitor/link/LinkInfo.java:
```diff
@@ -0,0 +1,91 @@
+
+        int percentSuccess = 100 * successChecks / getCheckCount();
+
+        if (percentSuccess >= LinkMonitorService.percentWorking) {
```
`percentWorking` may be better named `percentWorkingThreshold` (if that's indeed what it is)

danfruehauf replied:
That can be renamed, not a biggie.

jkburges replied:
Like you say, no bigger. `percentWorking` implies (to me, at least) it's holding the actual value rather than the threshold.

jkburges commented on web/src/main/java/org/fao/geonet/monitor/link/LinkMonitorInterface.java:
```diff
@@ -0,0 +1,44 @@
+import jeeves.server.resources.ResourceManager;
+import org.fao.geonet.GeonetContext;
+
+public class LinkMonitorInterface {
```
Seems a bit misleading to have a `class` named `...Interface`.  Just sayin'.

danfruehauf replied:
Umn, OK, fair enough. Would you like it as `LinkMonitorAbstractClass`? `LinkMonitorBase`? I'm not so sure.

jkburges replied:
I dunno about this one.  It's *probably* ok,  but if it's actually got implementation in it, then I think `LinkMonitorAbstract` would be preferable.  (think I said `AbstractLinkMonitor` above, but your name is more consistent).

jkburges commented on web/src/main/java/org/fao/geonet/monitor/link/LinkMonitorInterface.java:
```diff
@@ -0,0 +1,44 @@
+    private static Logger logger;
+    private static long reindexInterval;
+
+    public LinkMonitorInterface() {
```
This can just be deleted, no?

danfruehauf replied:
True.

jkburges commented on web/src/main/java/org/fao/geonet/monitor/link/LinkMonitorInterface.java:
```diff
@@ -0,0 +1,44 @@
+import org.fao.geonet.GeonetContext;
+
+public class LinkMonitorInterface {
+    private static ResourceManager resourceManager;
```
Should these be static (except perhaps for the logger)?  Gonna get some funny stuff happening if we have more than one `LinkMonitorInterface`.

danfruehauf replied:
You should not have more than one of that.

jkburges replied:
Ok.  The logger should be static too by the way. Not sure why I thought that!

jkburges commented on web/src/main/java/org/fao/geonet/monitor/link/LinkMonitorInterface.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    public void check() {}
+}
```
end of line

danfruehauf replied:
Bloody intellij.

jkburges replied:
Lolz

jkburges commented on web/src/main/java/org/fao/geonet/monitor/link/LinkMonitorInterface.java:
```diff
@@ -0,0 +1,44 @@
+    public void init(ResourceManager resourceManager, GeonetContext geonetContext, Logger logger, long reindexInterval) {
+        this.resourceManager = resourceManager;
+        this.geonetContext = geonetContext;
+        this.logger = logger;
```
Usual for classes to have their own loggers, as mentioned previously.

jkburges commented on web/src/main/java/org/fao/geonet/monitor/link/LinkMonitorService.java:
```diff
@@ -0,0 +1,142 @@
+
+public class LinkMonitorService extends LinkMonitorInterface {
+
+    public enum Status {
```
This enum isn't used in this class.  Would it be better off somewhere else (`LinkInfo` perhaps)?

jkburges commented on web/src/main/java/org/fao/geonet/monitor/link/LinkMonitorService.java:
```diff
@@ -0,0 +1,142 @@
+        UNKNOWN
+    }
+
+    public static int percentWorking = 90;
```
Does all this stuff need to be configurable?

danfruehauf replied:
Yeah, something I wrote in the email this morning, so... Yes, will be moved to config.

jkburges replied:
:+1:

jkburges commented on web/src/main/java/org/fao/geonet/monitor/link/LinkMonitorService.java:
```diff
@@ -0,0 +1,142 @@
+        try {
+            return getAllRecordsLucene();
+        } catch (Exception e) {
+            e.printStackTrace();
```
log rather than print to stderr.

jkburges commented on web/src/main/java/org/fao/geonet/monitor/link/LinkMonitorService.java:
```diff
@@ -0,0 +1,142 @@
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+        return null;
```
The result of this method is used (in `reindex()`) without a null check.  Therefore, a NPE lies in wait.

Under what circumstances can `getAllRecordsLucene()` throw an exception?  Maybe better for this method to return an empty list in this case?

jkburges commented on web/src/main/java/org/fao/geonet/monitor/link/LinkMonitorService.java:
```diff
@@ -0,0 +1,142 @@
+                }
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
```
log

jkburges commented on web/src/main/java/org/fao/geonet/monitor/link/LinkMonitorService.java:
```diff
@@ -0,0 +1,142 @@
+    }
+
+    private boolean needReindex() {
+        long now = System.currentTimeMillis() / 1000l;
```
Seems to be a lot of conversion to seconds going on all over the place - why not just deal in ms?  That's typical (although not *always* the case) when dealing in timestamps as longs, anyway.

jkburges commented on web/src/main/java/org/fao/geonet/monitor/link/LinkMonitorService.java:
```diff
@@ -0,0 +1,142 @@
+            getLogger().debug(String.format("Checking record '%s'", record.getKey()));
+            record.getValue().check();
+            try {
+                Thread.sleep(betweenChecksIntervalMs);
```
Explain why we've got the check interval? Is that so that we don't flog the endpoint?

danfruehauf replied:
I commented on that variable up in the class, I think it'd be good to have it configurable. Otherwise we can potentially hammer services on the same server - which may or may not result in the remote end blocking us...

jkburges replied:
Yep, sorry, who reads comments?!

jkburges commented on web/src/main/java/org/fao/geonet/monitor/link/LinkMonitorService.java:
```diff
@@ -0,0 +1,142 @@
+            try {
+                Thread.sleep(betweenChecksIntervalMs);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
```
log

jkburges commented on web/src/main/java/org/fao/geonet/monitor/link/LinkMonitorService.java:
```diff
@@ -0,0 +1,142 @@
+            return true;
+        }
+    }
+}
```
EOL

jkburges commented:
I'm possibly about to run out of time reviewing this today - will continue on with it tomorrow if so.

jkburges commented on web/src/main/java/org/fao/geonet/monitor/link/MetadataRecordInfo.java:
```diff
@@ -0,0 +1,147 @@
+
+    private void ReportStatusChange(LinkMonitorService.Status prevStatus, LinkMonitorService.Status newStatus) {
+        if (prevStatus == newStatus)
+            return;
```
braces?

jkburges commented on web/src/test/java/org/fao/geonet/monitor/link/MetadataRecordInfoTest.java:
```diff
@@ -0,0 +1,48 @@
+        statusList.add(LinkMonitorService.Status.FAILED);
+        assertEquals(LinkMonitorService.Status.FAILED, MetadataRecordInfo.getStatus(statusList));
+    }
+}
```
EOL

jkburges commented:
Ok, I've finished my first pass at reviewing this extensive body of work.  That would have to be my overriding general comment - that there is a ton of code here for something that perhaps isn't that complex, and I know that that is more a result of it executing within the context of geonetwork than anything else.

It might be worth getting someone who knows geonetwork to review this also, because they're going to do a much better job than me in terms of reviewing the overall design and style.

danfruehauf commented:
@jkburges Thanks for going through this. I am going to fix most of those things (if not all).

I am not super keen on separating the checkers into a base class and then WMS and WFS, it reduces clarity, but the rest of the stuff I'm taking.

jkburges commented on web/src/main/java/org/fao/geonet/monitor/link/LinkInfo.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    public void check() {
+        checkInfoList.add(new CheckInfo(onlineResource, linkChecker));
```
May as well call `addCheckInfo(new CheckInfo(...))`.

danfruehauf commented:
@jkburges I think I'm done modifying those things. I included also the state memoization because it was a relatively simple change and not intrusive. I also refactored the LinkChecker* interfaces and refactored them so `toString()` and such do not take parameters. Altogether it's too bad Java 7 does not support static functions in interfaces, but oh well. I eliminated quite a bit of code around there and it looks better IMO.

I am going to battle test this now, so please don't merge yet until I give you the green light.

jkburges commented:
:+1: apart from [the one small comment](https://github.com/aodn/core-geonetwork/commit/30193312861ba7c97df5abd454e3d0f577f4d8e7#commitcomment-13405737).

I'm on kid pick up today, leaving about 2.30pm, so I'm happy for you to merge this @danfruehauf when you're ready.
